001-runtime-interface/proposal.md

```diff
@@ -0,0 +1,191 @@
+caches.
+
+Depending on the kind of step (or check) being run, Core can create the required input, output and
+cache artifacts using `CreateArtifact`. The `Artifact`s created can then be included as part of the
+`RunnableSpec` for the next `Runnable`. Consider the following pipeline snippet as an example:
+
+pipeline.yml:
+```
+jobs:
+- name: unit
+  plan:
+  - get: booklit
+    trigger: true
+  - task: test
+    file: booklit/ci/test.yml
+```
+
+test.yml:
+```
+---
+platform: linux
+
+image_resource:
+  type: registry-image
+  source: {repository: golang}
+
+inputs:
+- name: booklit
+
+run:
+  path: booklit/ci/test
+```
+
+When setting up the build plan, the Concourse Core knows that the output of the `get` step is an 
+input to the `task` step. Step execution would be as follows:
+
+1) `storage.CreateArtifact()` to represent "booklit"
+2) `runtime.Run()` the `get` step with the "booklit" `Artifact` declared as an output in the
+   `RunnableSpec`
+3) `runtime.Run()` the `task` step with the "booklit" `Artifact` as an input in the `RunnableSpec`
+4) `storage.DestroyArtifact()` "booklit" when no longer required
+
+Note that in this case, Core does not care about the actual contents of the `Artifact`; it assumes
+that if the `get` was successful, the `Artifact` is correctly populated.
+
+In the current runtime, `Storage` would be modeled as a distributed "blobstore" on different
+workers (using Baggageclaim). `CreateArtifact` could simply create a database artifact object
+that isn't backed by any actual storage until one of two things occur:
+1) `artifact.Store()` is called
+2) the `Artifact` is passed in as an input/output for a `Runnable`
+
+One implication of this interface is that Core would need to manage the lifecycle of an `Artifact`.
+Note that this does not mean that Core would need to manage the volume lifecycle; instead, the
+volume lifecycle would depend on the artifact so that once the `Artifact` is deleted, any associated
+volumes would also get deleted.
+
+## Artifact Interface
+
+In most cases, the Concourse Core does not need to know the actual contents of an `Artifact`; the
+contents are only really used when the artifact is attached to a `Runnable`. However, there are a
+few cases where code that lies above the `Run/Store` abstraction needs to read the contents of an
+artifact. They are:
+
+1) *Running `fly execute`:* Users may upload or download user artifacts from their local machine.
+Since these `Artifact`s need to be streamed to/from `fly`, their contents need to be exposed to
+the API.
+2) *Fetching images:* Because Concourse allows the user to specify an `image_resource` can be
+specified in the task definition as opposed to the pipeline, there exists a use case where a user
+may first do a `get` and then run a `task` that was defined in the output `Artifact` of the `get`.
+
+To satisfy these use cases, we need to `Store()` and `Retrieve()` information from within an
```

Comment:
Is Retrieve required ? Specially because we're indicating that Runnable would be able to populate inputs from the Storage interface. Unless Runnable would consume a list of artifacts and use `Retrieve`

Reply:
@xtreme-sameer-vohra Retrieve is required for `fly execute`. If you `fly execute` a task with outputs, the outputs would need to be downloaded to your local computer. The API would need to call Retrieve to do this.